partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnModulo.kt

```diff
@@ -4,126 +4,78 @@
-    ),
-
-) { args ->
-    val arg0 = args[0].bigDecimal
-    val arg1 = args[1].bigDecimal
-    Datum.decimal(arg0 % arg1)
-}
-
-internal val Fn_MODULO__FLOAT32_FLOAT32__FLOAT32 = Function.static(
-
-    name = "modulo",
-    returns = PType.real(),
-    parameters = arrayOf(
-        Parameter("lhs", PType.real()),
-        Parameter("rhs", PType.real()),
-    ),
-
-) { args ->
-    val arg0 = args[0].float
-    val arg1 = args[1].float
-    Datum.real(arg0 % arg1)
-}
-
-internal val Fn_MODULO__FLOAT64_FLOAT64__FLOAT64 = Function.static(
-
-    name = "modulo",
-    returns = PType.doublePrecision(),
-    parameters = arrayOf(
-        Parameter("lhs", PType.doublePrecision()),
-        Parameter("rhs", PType.doublePrecision()),
-    ),
-
-) { args ->
-    val arg0 = args[0].double
-    val arg1 = args[1].double
-    Datum.doublePrecision(arg0 % arg1)
+internal object FnModulo : DiadicArithmeticOperator("modulo") {
+
+    init {
+        fillTable()
+    }
+
+    override fun getTinyIntInstance(tinyIntLhs: PType, tinyIntRhs: PType): Function.Instance {
+        return basic(PType.tinyint()) { args ->
+            @Suppress("DEPRECATION") val arg0 = args[0].byte
+            @Suppress("DEPRECATION") val arg1 = args[1].byte
+            Datum.tinyint((arg0 % arg1).toByte())
+        }
+    }
+
+    override fun getSmallIntInstance(smallIntLhs: PType, smallIntRhs: PType): Function.Instance {
+        return basic(PType.smallint()) { args ->
+            val arg0 = args[0].short
+            val arg1 = args[1].short
+            Datum.smallint((arg0 % arg1).toShort())
+        }
+    }
+
+    override fun getIntegerInstance(integerLhs: PType, integerRhs: PType): Function.Instance {
+        return basic(PType.integer()) { args ->
+            val arg0 = args[0].int
+            val arg1 = args[1].int
+            Datum.integer(arg0 % arg1)
+        }
+    }
+
+    override fun getBigIntInstance(bigIntLhs: PType, bigIntRhs: PType): Function.Instance {
+        return basic(PType.bigint()) { args ->
+            val arg0 = args[0].long
+            val arg1 = args[1].long
+            Datum.bigint(arg0 % arg1)
+        }
+    }
+
+    override fun getNumericInstance(numericLhs: PType, numericRhs: PType): Function.Instance {
+        return basic(PType.numeric()) { args ->
+            val arg0 = args[0].bigInteger
+            val arg1 = args[1].bigInteger
+            Datum.numeric(arg0 % arg1)
+        }
+    }
+
+    override fun getDecimalInstance(decimalLhs: PType, decimalRhs: PType): Function.Instance {
+        val p = decimalLhs.precision - decimalLhs.scale + decimalRhs.scale + Math.max(6, decimalLhs.scale + decimalRhs.precision + 1)
+        val s = Math.max(6, decimalLhs.scale + decimalRhs.precision + 1)
+        return basic(PType.decimal()) { args ->
+            val arg0 = args[0].bigDecimal
+            val arg1 = args[1].bigDecimal
+            Datum.decimal(arg0 % arg1, p, s)
+        }
+    }
```

Comment:
The computation for precision and scale is still following SQL Server's division rather than modulo. Should be 

p = min(p1 - s1, p2 - s2) + max(s1, s2)
s = max(s1, s2)

Reply:
Also would be helpful for future maintenance to include a comment on the precision/scale computation

Reply:
Added!